packages/angular/build/src/tools/esbuild/angular/compilation/aot-compilation.ts

```diff
@@ -245,11 +269,37 @@ export class AotCompilation extends AngularCompilation {
         continue;
       }
 
-      if (angularCompiler.incrementalCompilation.safeToSkipEmit(sourceFile)) {
+      if (
+        angularCompiler.incrementalCompilation.safeToSkipEmit(sourceFile) &&
+        !affectedFiles.has(sourceFile)
+      ) {
+        continue;
+      }
+
+      if (useTypeScriptTranspilation) {
+        typeScriptProgram.emit(sourceFile, writeFileCallback, undefined, undefined, transformers);
         continue;
       }
 
-      typeScriptProgram.emit(sourceFile, writeFileCallback, undefined, undefined, transformers);
+      // When not using TypeScript transpilation, directly apply only Angular specific transformations
+      const transformResult = ts.transform(
```

Comment:
It would be nice if we'd know here if the compiler has any transforms for this source file pending, otherwise all of this work is essentially a noop. This is known in `TraitCompiler` (its `filesWithoutTraits` field, or inversely when the source file is not present in `fileToClasses`).

I also wonder how using the printer affects source mapping, as esbuild will now map into transformed TS and there's no mapping back to the original source file, as the printer is incapable of generating source maps TMK.

Reply:
Definitely opportunities to further enhance with some potential adjustments to the AOT compiler.

Sourcemaps are the main reason this is still draft.  First stage will probably be limiting this to when sourcemaps are disabled which is most production builds. Production builds are also where this change provides the most benefit.
After that, exploring sourcemap production from the TS side would be the next step.  TS does export the needed elements (`createSourceMapGenerator`, for instance) from the JS but are not present in the typings.